docs/src/main/mdoc/consumers.md

```diff
@@ -0,0 +1,314 @@
+
+The following imports are assumed throughout this page.
+
+```scala mdoc:silent
+import cats.effect._
+import cats.implicits._
+import fs2.kafka._
+import scala.concurrent.duration._
+```
+
+## Deserializers
+
+[`Deserializer`][deserializer] describes functional composable deserializers for record keys and values. We generally require two deserializers: one for the record key and one for the record value. Deserializers are provided implicitly for many standard library types, including:
+
+- `Array[Byte]`, `Double`, `Float`, `Int`, `Long`, `Short`, `String`, and `UUID`.
+
+There are also deserializers for types which carry special meaning:
+
+- `Option[A]` to deserialize occurrances of `null` as `None`, and
+
+- `Unit` to ignore the serialized bytes and always use `()`.
+
+For more involved types, we need to resort to custom deserializers.
+
+### Custom Deserializers
+
+`Deserializer[F[_], A]` describes a function `Array[Byte] => F[A]`, while also having access to the topic name and record [`Headers`][headers]. There are many [functions][deserializer$] available for creating custom deserializers, with the most basic one being `instance`, which simply creates a deserializer from a provided function.
+
+```scala mdoc:silent
+Deserializer.instance {
+  (topic, headers, bytes) =>
+    IO.pure(bytes.dropWhile(_ == 0))
+}
+```
+
+If the deserializer only needs access to the bytes, like in the case above, use `lift`.
+
+```scala mdoc:silent
+Deserializer.lift(bytes => IO.pure(bytes.dropWhile(_ == 0)))
+```
+
+To support different deserializers for different topics, use `topic`.
+
+```scala mdoc:silent
+Deserializer.topic[IO, String] {
+  case "first"  => Deserializer[IO, String]
+  case "second" => Deserializer[IO, Int].map(_.toString)
+}
+```
+
+For unmatched topics, an [`UnexpectedTopicException`][unexpectedtopicexception] is raised.
+
+Use `headers` for different deserializers depending on record headers.
+
+```scala mdoc:silent
+Deserializer.headers[IO, String] { headers =>
+  headers("format").map(_.as[String]) match {
+    case Some("string") => Deserializer[IO, String]
+    case Some("int")    => Deserializer[IO, Int].map(_.toString)
+    case Some(format)   => Deserializer.failWith(s"unknown format: $format")
+    case None           => Deserializer.failWith("format header is missing")
+  }
+}
+```
+
+In the example above, `failWith` raises a [`SerializationException`][serializationexception] with the provided message.
+
+### Java Interoperability
+
+If we have a Java Kafka deserializer, use `delegate` to create a `Deserializer`.
+
+```scala mdoc:silent
+Deserializer.delegate[IO, String] {
+  new KafkaDeserializer[String] {
+    def deserialize(topic: String, data: Array[Byte]): String = new String(data)
+    def configure(configs: java.util.Map[String, _], isKey: Boolean): Unit = ()
+    def close(): Unit = ()
+  }
+}
+```
+
+If the deserializer performs _side effects_, follow with `suspend` to capture them properly.
```

Comment:
What will happen if we don't follow with `suspend` event though the serializer performs _side effects_?

Reply:
It should mainly impact errors, i.e. depending on the effect type you use, errors will either be handled anyway (with e.g. `IO` and `Task`) or treated as bug (with e.g. `ZIO`), resulting in immediate shutdown.

It's the same situation as `IO.pure(throw new RuntimeException)`, except we're in a `flatMap`-chain, meaning `IO` and `Task` will catch the error anyway, while `ZIO` won't. I'd like us to not describe how deserializers are used for the user though, so will leave it for now.

Reply:
sounds good to me!!